### Task 1
Data Cleaning Challenge on Sales Data.

In [6]:
import pandas as pd
sales_2017=pd.read_csv("Sales-Transactions-2017.csv.csv")
sales_2018=pd.read_csv("Sales-Transactions-2018.csv.csv")
sales_2019=pd.read_csv("Sales-Transactions-2019.csv.csv")
sales_edited=pd.read_csv("Sales-Transactions-Edited.csv.csv")
sales_fulldata=pd.concat(objs=[sales_2017,sales_2018,sales_2019])

### Observing cleaned and uncleaned data shape

In [7]:
sales_edited.shape

(95562, 6)

In [8]:
sales_fulldata.shape

(111206, 9)

### Finding how many null values and dtypes in cleaned and non cleaned data

In [8]:
sales_edited.isna().sum()

Date       0
Voucher    0
Party      0
Product    0
Qty        0
Rate       1
dtype: int64

In [20]:
sales_fulldata.isna().sum()

Date               12591
Voucher            12557
Party                 40
Product            12591
Qty                12557
Rate               12558
Gross              12558
Disc              105609
Voucher Amount     83646
dtype: int64

In [9]:
sales_edited.dtypes

Date        object
Voucher      int64
Party       object
Product     object
Qty          int64
Rate       float64
dtype: object

In [10]:
sales_fulldata.dtypes

Date              object
Voucher           object
Party             object
Product           object
Qty               object
Rate              object
Gross             object
Disc              object
Voucher Amount    object
dtype: object

### Finding any duplicated values are there in full data

In [7]:
sales_fulldata[sales_fulldata.duplicated()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
14,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
17,,,dummy entry,,,,,,
39,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
41,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
43,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
19143,NaN,NaN,ET,NaN,NaN,NaN,NaN,NaN,NaN
19162,,,vinayaka sweets-kot,,,,,,
19163,,,hagunta,,,,,,
19172,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Duplicate data have null values also so its better to remove null values and then come to duplicate values

### Checking the few rows of the uncleaned data

In [12]:
sales_fulldata.head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
5,1/4/2017,Sal:2,SARNESWARA TRADERS,CLASSIC ENJOY(750),1,"1,320.00","1,320.00",NaN,NaN
6,1/4/2017,Sal:898,Lock,Vishnu 250ml,100,30,"3,000.00",100,"5,400.00"
7,1/4/2017,Sal:898,Lock,BLACK DOG-350ML,100,26,"2,600.00",100,NaN
8,,,khader vali late en,,,,,,
9,,,try,,,,,,


Observation many have party, product and date as null which seems like irrelevant data and will be removed

### Null values of party, product and date will be removed as these data are not meaningful 

In [29]:
sales_fulldata[(sales_fulldata["Party"].isna()) & (sales_fulldata["Date"].isna()) & (sales_fulldata["Product"].isna())]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
4286,NaN,Total,NaN,NaN,"725,098.20","892,995.90","10,156,375.86","99,183.00","10,079,469.86"
9440,NaN,Total,NaN,NaN,"864,048.00","842,990.34","11,738,640.00","76,449.00","11,690,343.00"
13637,NaN,Total,NaN,NaN,"742,449.00","639,799.63","9,994,892.00","115,371.00","9,905,775.00"
17195,NaN,Total,NaN,NaN,"569,166.00","684,525.17","9,196,270.50","68,314.00","9,147,636.50"
21653,NaN,Total,NaN,NaN,"800,911.60","916,383.98","11,487,277.00","80,427.00","11,434,171.00"
25642,NaN,Total,NaN,NaN,"616,175.60","654,104.12","9,814,351.46","88,940.00","9,749,457.46"
29365,NaN,Total,NaN,NaN,"638,342.70","621,104.41","9,496,760.91","65,056.00","9,453,182.91"
33121,NaN,Total,NaN,NaN,"551,198.00","643,446.58","9,132,584.78","52,087.00","9,100,067.78"
37214,NaN,Total,NaN,NaN,"538,900.90","728,887.38","9,390,510.25","82,913.00","9,326,772.25"
40415,NaN,Total,NaN,NaN,"458,230.80","554,312.00","7,844,514.06","49,961.00","7,811,481.06"


In [9]:
sales_fulldata.dropna(subset=["Party","Date","Product"],how="all",inplace=True)

In [10]:
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


111206 - 111166 = Total 40 null values of party, product and date were removed

Still there are some values in sales_fulldata for date, product,qunatity, rate,gross 

In [15]:
sales_fulldata.isna().sum()

Date               12551
Voucher            12551
Party                  0
Product            12551
Qty                12551
Rate               12552
Gross              12552
Disc              105603
Voucher Amount     83640
dtype: int64

In [22]:
sales_fulldata[(sales_fulldata["Date"].isna()) & (sales_fulldata["Product"].isna()) & (sales_fulldata["Qty"].isna())]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
11,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
14,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
39,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
41,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
43,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
19093,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
19095,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
19097,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
19141,NaN,NaN,ET,NaN,NaN,NaN,NaN,NaN,NaN


In [11]:
sales_fulldata.dropna(subset=["Date","Product","Qty"],how="all",inplace=True)

In [24]:
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


### Again checking null values

In [25]:
sales_fulldata.isna().sum() #no null values

Date                  0
Voucher               0
Party                 0
Product               0
Qty                   0
Rate                  1
Gross                 1
Disc              93052
Voucher Amount    71089
dtype: int64

### Checking the data again for rows if any other data can be cleaned

In [28]:
sales_fulldata.head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
5,1/4/2017,Sal:2,SARNESWARA TRADERS,CLASSIC ENJOY(750),1,"1,320.00","1,320.00",NaN,NaN
6,1/4/2017,Sal:898,Lock,Vishnu 250ml,100,30,"3,000.00",100,"5,400.00"
7,1/4/2017,Sal:898,Lock,BLACK DOG-350ML,100,26,"2,600.00",100,NaN
8,,,khader vali late en,,,,,,
9,,,try,,,,,,


Some records are having balank or white space these data will be replaced with "Nan" values

### Records having balank or white space will be replaced with "Nan" values

In [13]:
columns_to_check = ["Date", "Voucher", "Product", "Qty", "Rate", "Gross", "Disc", "Voucher Amount"]

In [41]:
sales_fulldata[(sales_fulldata[columns_to_check] == " ").all(axis=1)]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
8,,,khader vali late en,,,,,,
9,,,try,,,,,,
12,,,DUMMY ENTRY,,,,,,
15,,,dummy entry,,,,,,
17,,,dummy entry,,,,,,
...,...,...,...,...,...,...,...,...,...
18949,,,credit bill,,,,,,
19052,,,8/10 late entry,,,,,,
19057,,,10-Jul,,,,,,
19162,,,vinayaka sweets-kot,,,,,,


Except party all have blank spaces for 3051 rows

### Replacing the white spaces to null, once replaced checking the null values

In [14]:
import numpy as np
sales_fulldata[columns_to_check] = sales_fulldata[columns_to_check].replace(r"^\s*$", np.nan, regex=True)


In [44]:
sales_fulldata.isna().sum()

Date               3053
Voucher            3053
Party                 0
Product            3053
Qty                3051
Rate               3052
Gross              3054
Disc              96105
Voucher Amount    74142
dtype: int64

### Removing the null values using dropna

In [15]:
sales_fulldata.dropna(subset=columns_to_check,how="all",inplace=True)
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


95564 rows columns achieved cleaned data had 95562 rows

In [53]:
sales_edited

,Date,Voucher,Party,Product,Qty,Rate
0,1/4/2017,1,SOLANKI PLASTICS,DONA-VAI-9100,2,1690.0
1,1/4/2017,1,SOLANKI PLASTICS,LITE FOAM(1200),6,1620.0
2,1/4/2017,2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23.0
3,1/4/2017,2,SARNESWARA TRADERS,LITE FOAM(1200),6,1620.0
4,1/4/2017,2,SARNESWARA TRADERS,DONA-VAI-9100,5,1690.0
...,...,...,...,...,...,...
95557,12/9/2019,4265,TP13,SPOON MED M.W,20,11.0
95558,12/9/2019,4266,K.SRIHARI,SMART BOUL(48),1,1830.0
95559,12/9/2019,4267,SMS,SMARTBOUL GLA(4000),1,1520.0
95560,12/9/2019,4268,ANILFANCY,RR WINEGLASS,100,20.0


### Deleting the columns "Gross","Disc", and "Voucher Amount" as it was not there in cleaned data?

In [16]:
Delete_column=["Gross","Disc","Voucher Amount"]
sales_fulldata.drop(columns=Delete_column,inplace=True)

In [69]:
sales_fulldata.dtypes

Date       object
Voucher    object
Party      object
Product    object
Qty        object
Rate       object
dtype: object

### Converting the datatypes of date (dateformat) , voucher (int64), quantity (int64), and rate (float64)?

In [17]:
sales_fulldata["Date"]=pd.to_datetime(sales_fulldata["Date"],format="%m/%d/%Y",errors="coerce")

### Vocucher as Sal: string which needs to be removed before converting it to int64?

In [18]:
sales_fulldata["Voucher"].replace(r"Sal:","",regex=True,inplace=True)


C:\Users\Q848957\AppData\Local\Temp\ipykernel_17700\4134258599.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sales_fulldata["Voucher"].replace(r"Sal:","",regex=True,inplace=True)


In [87]:
sales_fulldata["Voucher"].astype("int64") #Nan values were there so getting the below error

ValueError: cannot convert float NaN to integer

### Found two Nan data in date voucher product which needs to be deleted?

In [19]:
sales_fulldata[sales_fulldata["Voucher"].isna()]
sales_fulldata.dropna(subset="Voucher",inplace=True)

In [20]:
sales_fulldata["Voucher"]=sales_fulldata["Voucher"].astype("int64")


In [117]:
sales_fulldata["Qty"].astype("int64") #not able to convert in to int64 on researching seems like need to remove comma

ValueError: invalid literal for int() with base 10: '1000.00'

In [21]:
sales_fulldata["Qty"].replace(r",","",regex=True,inplace=True)
sales_fulldata["Qty"].astype("int64") #still not working seems like some data are in float 
sales_fulldata["Qty"].round().astype("int64")

C:\Users\Q848957\AppData\Local\Temp\ipykernel_17700\4121309962.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sales_fulldata["Qty"].replace(r",","",regex=True,inplace=True)


ValueError: invalid literal for int() with base 10: '1000.00'

In [22]:
sales_fulldata["Qty"]=sales_fulldata["Qty"].astype("float64").round()
sales_fulldata["Qty"]=sales_fulldata["Qty"].astype("int64")

In [23]:

sales_fulldata["Rate"].replace(r",","",regex=True,inplace=True)
sales_fulldata["Rate"]=sales_fulldata["Rate"].astype("float64")

C:\Users\Q848957\AppData\Local\Temp\ipykernel_17700\4277905212.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sales_fulldata["Rate"].replace(r",","",regex=True,inplace=True)


In [24]:
sales_fulldata.dtypes

Date       datetime64[ns]
Voucher             int64
Party              object
Product            object
Qty                 int64
Rate              float64
dtype: object

In [142]:
sales_fulldata.shape

(95562, 6)

In [25]:
sales_fulldata = sales_fulldata.reset_index(drop=True) # needed to do this as the index numbers were jumbling

In [148]:
sales_fulldata

,Date,Voucher,Party,Product,Qty,Rate
0,2017-01-04,1,SOLANKI PLASTICS,DONA-VAI-9100,2,1690.0
1,2017-01-04,1,SOLANKI PLASTICS,LITE FOAM(1200),6,1620.0
2,2017-01-04,2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23.0
3,2017-01-04,2,SARNESWARA TRADERS,LITE FOAM(1200),6,1620.0
4,2017-01-04,2,SARNESWARA TRADERS,DONA-VAI-9100,5,1690.0
...,...,...,...,...,...,...
95557,2019-10-10,4935,K.SRIHARI,16*20(100-W),140,26.0
95558,2019-10-10,4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4
95559,2019-10-10,4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16.0
95560,2019-10-10,4935,K.SRIHARI,10*12 RK,800,8.5


In [149]:
sales_edited

,Date,Voucher,Party,Product,Qty,Rate
0,1/4/2017,1,SOLANKI PLASTICS,DONA-VAI-9100,2,1690.0
1,1/4/2017,1,SOLANKI PLASTICS,LITE FOAM(1200),6,1620.0
2,1/4/2017,2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23.0
3,1/4/2017,2,SARNESWARA TRADERS,LITE FOAM(1200),6,1620.0
4,1/4/2017,2,SARNESWARA TRADERS,DONA-VAI-9100,5,1690.0
...,...,...,...,...,...,...
95557,12/9/2019,4265,TP13,SPOON MED M.W,20,11.0
95558,12/9/2019,4266,K.SRIHARI,SMART BOUL(48),1,1830.0
95559,12/9/2019,4267,SMS,SMARTBOUL GLA(4000),1,1520.0
95560,12/9/2019,4268,ANILFANCY,RR WINEGLASS,100,20.0


### Checking duplicate values after removing null?

In [ ]:
Not removing duplicate as it will be removed last

In [26]:
sales_fulldata[sales_fulldata.duplicated()]

,Date,Voucher,Party,Product,Qty,Rate
1335,2017-12-04,439,TP13,Classic 250ml,100,23.0
1873,NaT,443,SHAIK SIFULLA,CYCLE-BK-13*16,10,13.0
1990,NaT,552,TP13,Vishnu 300ml,100,27.5
2041,NaT,497,Lakshmifancy,"TAPE ROLL 2"" 50MTRS",4,125.0
2485,NaT,599,SPM PROVISION-DAKKI,NO-00,1,150.0
...,...,...,...,...,...,...
94481,NaT,4608,SEKHAR MARKET,14 SILVER PLATES,60,260.0
94901,2019-01-10,4795,TP13,NAPKEEN 2ND,140,8.5
95241,2019-07-10,4844,VENKATESWARA GENERA,NO-0,5,160.0
95272,2019-07-10,4896,TP13,Vishnu 250ml,100,29.0


In [27]:
sales_edited[sales_edited.duplicated()]

,Date,Voucher,Party,Product,Qty,Rate
411,1/8/2017,3841,TP13,PP,5,135.0
1080,2/6/2017,2099,TP13,9*12 STAR-DCUT,5,100.0
2171,3/5/2017,1122,TP13,CLASSIC 250ML,100,23.0
2176,3/5/2017,1122,TP13,CLASSIC 250ML,100,23.0
2331,3/6/2017,2107,TP13,CLASSIC 300ML,100,24.5
...,...,...,...,...,...,...
94125,11/3/2019,8986,SURI,24*30 COVER,50,75.0
94565,11/7/2019,2589,TP13,RUBBER,5,290.0
94786,12/1/2019,7564,TP13,VISHNU 300ML,30,28.0
95064,12/3/2019,9060,TP13,DURGA 13*16 BLUE,200,11.3


### Task 1 completed

### Task 2 : Merging

### Merging function example

In [28]:
import pandas as pd
df_employees = pd.DataFrame({
    'Employee_ID': [101, 102, 103, 104],
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Department': ['HR', 'IT', 'Finance', 'IT']
})

df_salaries = pd.DataFrame({
    'Employee_ID': [101, 102, 103, 105],
    'Salary': [70000, 85000, 65000, 90000]
})

### Merge inner join
Only matching rows

In [36]:
df_inner=df_employees.merge(df_salaries,how="inner")
df_inner

,Employee_ID,Name,Department,Salary
0,101,Alice,HR,70000
1,102,Bob,IT,85000
2,103,Charlie,Finance,65000


### Merge outer join
Matching and unmatching rows

In [38]:
df_outer=df_employees.merge(df_salaries,on="Employee_ID",how="outer")
df_outer

,Employee_ID,Name,Department,Salary
0,101,Alice,HR,70000.0
1,102,Bob,IT,85000.0
2,103,Charlie,Finance,65000.0
3,104,David,IT,NaN
4,105,NaN,NaN,90000.0


### Left join
All left rows + matching rows on the right

In [39]:
df_left=df_employees.merge(df_salaries,on="Employee_ID",how="left")
df_left

,Employee_ID,Name,Department,Salary
0,101,Alice,HR,70000.0
1,102,Bob,IT,85000.0
2,103,Charlie,Finance,65000.0
3,104,David,IT,NaN


### Right Join
All right rows + matching rows on the left

In [46]:
df_right=pd.DataFrame.merge(df_employees,df_salaries,on="Employee_ID",how="right")
df_right

,Employee_ID,Name,Department,Salary
0,101,Alice,HR,70000
1,102,Bob,IT,85000
2,103,Charlie,Finance,65000
3,105,NaN,NaN,90000


In [53]:
df_employees_02 = pd.DataFrame({
    'Employee_ID': [101, 102, 103, 104],
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Department': ['HR', 'IT', 'Finance', 'IT']
})

df_salaries_02 = pd.DataFrame({
    'E_ID': [101, 102, 103, 105],
    'Salary': [70000, 85000, 65000, 90000]
})

df_names_02 = pd.DataFrame({
    'Employee_ID': [101, 106, 103, 108],
    'Name': ['Alice', 'king', 'Charlie', 'Russal'],
    'Department': ['HR', 'IT', 'Finance', 'IT']
})

### Using left on and right on to do outer join

In [50]:
df_merged_outer=pd.merge(df_employees_02,df_salaries_02,left_on="Employee_ID",right_on="E_ID",how="outer")
df_merged_outer["Employee_ID"]=df_merged_outer["Employee_ID"].fillna(df_merged_outer["E_ID"]) # trying to comnine Eid with Employee id
df_merged_outer.drop("E_ID",axis=1,inplace=True)
df_merged_outer

,Employee_ID,Name,Department,Salary
0,101.0,Alice,HR,70000.0
1,102.0,Bob,IT,85000.0
2,103.0,Charlie,Finance,65000.0
3,104.0,David,IT,NaN
4,105.0,NaN,NaN,90000.0


### Suffix parameter in merge function

In [55]:
df_merged_suffix=pd.merge(df_employees_02,df_names_02,on="Employee_ID",how="outer")
df_merged_suffix

,Employee_ID,Name_x,Department_x,Name_y,Department_y
0,101,Alice,HR,Alice,HR
1,102,Bob,IT,NaN,NaN
2,103,Charlie,Finance,Charlie,Finance
3,104,David,IT,NaN,NaN
4,106,NaN,NaN,king,IT
5,108,NaN,NaN,Russal,IT


In [58]:
df_merged_suffix=pd.merge(df_employees_02,df_names_02,on="Employee_ID",how="outer",suffixes=("First","Last"))
df_merged_suffix

,Employee_ID,NameFirst,DepartmentFirst,NameLast,DepartmentLast
0,101,Alice,HR,Alice,HR
1,102,Bob,IT,NaN,NaN
2,103,Charlie,Finance,Charlie,Finance
3,104,David,IT,NaN,NaN
4,106,NaN,NaN,king,IT
5,108,NaN,NaN,Russal,IT


### Task 2 completed

### Task 3

### List Comprehension 2 examples
List comprehension in Python is a compact syntax used to construct new lists from existing iterables.


for item in iterable:
    if condition:
        expression     is equal to list comprehension         [expression for item in iterable if condition]

### Printing first 20 even numbers without list expression

In [59]:
for i in range(1,20):
    if i%2==0:
        print(i)

2
4
6
8
10
12
14
16
18


### Print first 20 even numbers with list expression

In [61]:
[print (i) for i in range(1,20) if i%2==0] # this one creates a list with none value as the print function returns none

2
4
6
8
10
12
14
16
18


[None, None, None, None, None, None, None, None, None]

In [63]:
[i for i in range(1, 20) if i % 2 == 0] #The brackets outside is what creates a list

[2, 4, 6, 8, 10, 12, 14, 16, 18]

In [64]:
print([i for i in range(1, 20) if i % 2 == 0]) #no use in printing

[2, 4, 6, 8, 10, 12, 14, 16, 18]


In [ ]:
Even_numbers=[i for i in range(1, 20) if i % 2 == 0]

### Data Transformation (Mapping)

**Squaring numbers**

In [65]:
[x**2 for x in range (10)] # without if condition also it works only with expression and for iterable items

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

**Changing string casing**

In [3]:
word="Welcome home"
[word.upper() for word in word ] # stores each letter

['W', 'E', 'L', 'C', 'O', 'M', 'E', ' ', 'H', 'O', 'M', 'E']

**Type casting data**

In [68]:
Numbers=["4","5","6","7"]
[int(Numbers) for Numbers in Numbers]

[4, 5, 6, 7]

**Extracting sub-elements**

In [8]:
[word for word in word]

['W', 'e', 'l', 'c', 'o', 'm', 'e', ' ', 'h', 'o', 'm', 'e']

In [9]:
[word[0] for word in word] # the above and below are same

['W', 'e', 'l', 'c', 'o', 'm', 'e', ' ', 'h', 'o', 'm', 'e']

In [18]:
New_words=["King","Horse","Bishop"]
First_letter=[new[0] for new in New_words]
Last_letter=[new[-1] for new in New_words]
First_letter
Last_letter

['g', 'e', 'p']

###  2. Data Filtering

Extracting odd numbers:

In [72]:
[i%2!=0 for i in range(20)] #this will give only the boolean values so the expression should return real value if condtion can be used

[False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True]

In [73]:
[i for i in range(20) if i%2!=0] #if true then it stores the numbe in the list

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

**Filtering by string content**

In [74]:
word="Welcome home"
[word for word in word if word=="e"]

['e', 'e', 'e']

**Removing empty strings**

In [77]:
string_listst=["av","","we"]    
[s for s in string_listst if s] #this works only if there is an empty string value, even if you have space between the quotes it wont consider it as empty

['av', 'we']

### Conditional Modifications (If-Else)

In [29]:
Letters=["You","","we","our","Their"] 
[Q if Q == "our" else "our" for Q in Letters] # to use else neeeds to be used before for iteration

[False, False, False, 'our', False]

In [83]:
scores=[100,20,30,45,60,85,25]
["Pass" if score>35 else "Fail" for score in scores]

['Pass', 'Fail', 'Fail', 'Pass', 'Pass', 'Pass', 'Fail']

### Nested Loops & Matrix Operations

In [97]:
Matrix2=[(x, y) for x in range(3) for y in range(3)]
Matrix2

[(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]

In [98]:
Matrix3=[[j for j in range(3)] for i in range(5)] 
Matrix3

[[0, 1, 2], [0, 1, 2], [0, 1, 2], [0, 1, 2], [0, 1, 2]]

In [92]:
Matrix3=[[j for j in range(4)] for i in range(5)]
Matrix3

[[0, 1, 2, 3], [0, 1, 2, 3], [0, 1, 2, 3], [0, 1, 2, 3], [0, 1, 2, 3]]

In [89]:
type(Matrix3)

list

### Printing random 5 numbers with list comprehension

In [101]:
import random
[random.randint(1,100) for i in range(5)]

[25, 82, 84, 87, 29]

### Using function in list comprehension

In [30]:
def square(x):
    return x**2
[square(i) for i in range(10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

### Task 3 completed

### Task4


### Lambda function 

### What is lambda function?

A lambda function is a small anonymous function. A lambda function can take any number of arguments, but can only have one expression.

### Simple lambda function example

In [ ]:


first_number=2
second_number=3
add=lambda x,y: x+y
add(first_number,second_number)

### To do with elements in list

In [40]:


first_list=[2,1]
second_list=[36,6]
add=list(map(lambda x,y: x+y,first_list,second_list)) # use either map, filter or reduce when using list
add

[38, 7]

### Lambda Function with 2 examples 

Importing two dataset titanic and health information

In [1]:
import seaborn as sns
Health_information=sns.load_dataset("healthexp")
Titanic=sns.load_dataset("titanic")
Health_information

,Year,Country,Spending_USD,Life_Expectancy
0,1970,Germany,252.311,70.6
1,1970,France,192.143,72.2
2,1970,Great Britain,123.993,71.9
3,1970,Japan,150.437,72.0
4,1970,USA,326.961,70.9
...,...,...,...,...
269,2020,Germany,6938.983,81.1
270,2020,France,5468.418,82.3
271,2020,Great Britain,5018.700,80.4
272,2020,Japan,4665.641,84.7


In [2]:
Titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


### Using lambda in Map function

Use map() when you want to transform every single item in a list into something new.

**Map Example 1 (titanic): Creating a customized column named "description" to include both passenger genders and passenger class**

In [13]:
passenger_pairs=list(zip(Titanic["sex"],Titanic["pclass"]))
Titanic["description"]=list(map(lambda x: f"A {x[0]} passenger traveling in {x[1]} class", passenger_pairs))
Titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone,description
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False,A male passenger traveling in 3 class
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False,A female passenger traveling in 1 class
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True,A female passenger traveling in 3 class
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False,A female passenger traveling in 1 class
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True,A male passenger traveling in 3 class
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True,A male passenger traveling in 2 class
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True,A female passenger traveling in 1 class
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False,A female passenger traveling in 3 class
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True,A male passenger traveling in 1 class


### MAp Example 2 (healthxp): Calculate the difference between Year and current year

In [112]:
Health_information["Year_difference"]=list(map(lambda x: 2026-x, Health_information["Year"]))
Health_information

,Year,Country,Spending_USD,Life_Expectancy,LE_percent,Age,Year_difference
0,1970,Germany,252.311,70.6,70.6 Years Old,70.6 Years Old,56
1,1970,France,192.143,72.2,72.2 Years Old,72.2 Years Old,56
2,1970,Great Britain,123.993,71.9,71.9 Years Old,71.9 Years Old,56
3,1970,Japan,150.437,72.0,72.0 Years Old,72.0 Years Old,56
4,1970,USA,326.961,70.9,70.9 Years Old,70.9 Years Old,56
...,...,...,...,...,...,...,...
269,2020,Germany,6938.983,81.1,81.1 Years Old,81.1 Years Old,6
270,2020,France,5468.418,82.3,82.3 Years Old,82.3 Years Old,6
271,2020,Great Britain,5018.700,80.4,80.4 Years Old,80.4 Years Old,6
272,2020,Japan,4665.641,84.7,84.7 Years Old,84.7 Years Old,6


### Using lambda in Filter function

### Filter Example 1 (titanic): Isolating adult male survivors

In [56]:
import pandas as pd
Titanic_alive=pd.DataFrame(filter(lambda x: x.sex=="male" and x.alive=="yes" and x.age>18,Titanic.itertuples(index=False)))
Titanic_alive

,survived,pclass,sex,age,sibsp,parch,fare,embarked,_8,who,adult_male,deck,embark_town,alive,alone,description
0,1,2,male,34.0,0,0,13.0000,S,Second,man,True,D,Southampton,yes,True,A male passenger traveling in 2 class
1,1,1,male,28.0,0,0,35.5000,S,First,man,True,A,Southampton,yes,True,A male passenger traveling in 1 class
2,1,3,male,32.0,0,0,56.4958,S,Third,man,True,NaN,Southampton,yes,True,A male passenger traveling in 3 class
3,1,3,male,29.0,0,0,9.5000,S,Third,man,True,NaN,Southampton,yes,True,A male passenger traveling in 3 class
4,1,1,male,23.0,0,1,63.3583,C,First,man,True,D,Cherbourg,yes,False,A male passenger traveling in 1 class
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
64,1,3,male,27.0,0,0,6.9750,S,Third,man,True,NaN,Southampton,yes,True,A male passenger traveling in 3 class
65,1,3,male,27.0,0,0,8.6625,S,Third,man,True,NaN,Southampton,yes,True,A male passenger traveling in 3 class
66,1,3,male,32.0,0,0,56.4958,S,Third,man,True,NaN,Southampton,yes,True,A male passenger traveling in 3 class
67,1,1,male,51.0,0,0,26.5500,S,First,man,True,E,Southampton,yes,True,A male passenger traveling in 1 class


### Filter Example 2 (healthxp): Filter the country USA with life expectancy greater than 60 years using lambda filter function

In [64]:
Life_expectancy_USA= pd.DataFrame(filter(lambda x: x.Country == "USA" and x.Life_Expectancy > 60.0, Health_information.itertuples(index=False)))
Life_expectancy_USA

,Year,Country,Spending_USD,Life_Expectancy
0,1970,USA,326.961,70.9
1,1971,USA,357.988,71.2
2,1972,USA,397.097,71.2
3,1973,USA,439.302,71.4
4,1974,USA,495.114,72.0
5,1975,USA,560.750,72.7
6,1976,USA,638.851,72.9
7,1977,USA,726.241,73.3
8,1978,USA,808.884,73.5
9,1979,USA,908.963,73.9


### Using lambda in Reduce function

In [69]:
from functools import reduce
list1=[1,2,3,4]
add=reduce(lambda x,y:x+y,list1)
add



10

### Reduce Example 1 : Find the Maximum Fare Paid by a Survivor (Titanic Dataset)

In [94]:
survivor_rows = Titanic[Titanic["alive"] == "yes"]
Survivor_max_expense=reduce(lambda x,y: x if x>y.fare else y.fare,survivor_rows.itertuples(index=False),0.0)
Survivor_max_expense

512.3292

In [88]:
Titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone,description
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False,A male passenger traveling in 3 class
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False,A female passenger traveling in 1 class
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True,A female passenger traveling in 3 class
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False,A female passenger traveling in 1 class
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True,A male passenger traveling in 3 class
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True,A male passenger traveling in 2 class
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True,A female passenger traveling in 1 class
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False,A female passenger traveling in 3 class
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True,A male passenger traveling in 1 class


### Reduce Example 2: Calculate the Cumulative Health Spending of Germany

In [104]:
Healthdata=Health_information.itertuples(index=False)
Germany_expense=reduce(lambda total,row: total + row.Spending_USD if row.Country=="Germany" else total,Healthdata,0.0)
# here the first total should always be the carrier of the int number
#if you use if condition then else also needs to be used
#0.0 is the starting 'total' balance
Germany_expense


133364.01

In [98]:
Health_information

,Year,Country,Spending_USD,Life_Expectancy
0,1970,Germany,252.311,70.6
1,1970,France,192.143,72.2
2,1970,Great Britain,123.993,71.9
3,1970,Japan,150.437,72.0
4,1970,USA,326.961,70.9
...,...,...,...,...
269,2020,Germany,6938.983,81.1
270,2020,France,5468.418,82.3
271,2020,Great Britain,5018.700,80.4
272,2020,Japan,4665.641,84.7
